e2e/nomostest/registryproviders/image.go

```diff
@@ -0,0 +1,116 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registryproviders
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"kpt.dev/configsync/e2e/nomostest/gitproviders"
+	"kpt.dev/configsync/e2e/nomostest/testshell"
+)
+
+// use an auto-incrementing index to create unique file names for tarballs
+var imageIndex int
+
+// NewImage creates a new OCIImage object and associated tarball using the provided
+// Repository. The contents of the git repository will be bundled into a tarball
+// at the artifactDir. The resulting OCIImage object can be pushed to a remote
+// registry using its Push method.
+func NewImage(artifactDir string, shell *testshell.TestShell, repository *gitproviders.Repository, provider RegistryProvider) (*OCIImage, error) {
+	commitHash, err := repository.Hash()
+	if err != nil {
+		return nil, fmt.Errorf("getting hash: %w", err)
+	}
+	branch, err := repository.CurrentBranch()
+	if err != nil {
+		return nil, fmt.Errorf("getting branch: %w", err)
+	}
+	// Use branch/hash for context and imageIndex to enforce file name uniqueness.
+	// This avoids file name collision even if the test builds an image twice with
+	// a dirty repo state.
+	localSourceTgzPath := filepath.Join(artifactDir,
+		fmt.Sprintf("%s-%s-%d.tgz", branch, commitHash, imageIndex))
+	imageIndex++
+	if _, err := shell.ExecWithDebug("tar", "-cvzf", localSourceTgzPath, repository.Root); err != nil {
+		return nil, fmt.Errorf("packaging image: %w", err)
+	}
+	image := &OCIImage{
+		localSourceTgzPath: localSourceTgzPath,
+		Name:               repository.Name,
+		syncURL:            provider.OCISyncURL(repository.Name),
+		branch:             branch,
+		shell:              shell,
+	}
+	return image, nil
+}
+
+// OCIImage represents an OCI image that is pushed to a remote registry by the
+// test scaffolding. It uses git references as version tags to enable straightforward
+// integration with the git e2e tooling and to mimic how a user might leverage
+// git and OCI.
+type OCIImage struct {
+	localSourceTgzPath string
+	syncURL            string
+	branch             string
+	Name               string
+	Digest             string
+	shell              *testshell.TestShell
+}
+
+// FloatingTag returns the floating tag that initially points to this image.
+// This tag is suitable for use on RSync object spec but not for interacting with
+// the image from the test suite.
+// This uses the git branch as a version, such that the tag will be updated
+// whenever a new image is pushed from the git branch. This enables syncing
+// to an OCI image similar to a git branch.
+func (o *OCIImage) FloatingTag() string {
```

Comment:
`BranchTag` maybe? The fact that it's floating is important but not descriptive of what it is.

If you really want it in the name, maybe `FloatingBranchTag`?

Reply:
Done